crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -98,11 +98,91 @@ pub(crate) struct Store {
     inner: Arc<DynCryptoStore>,
     verification_machine: VerificationMachine,
     tracked_users_cache: Arc<DashSet<OwnedUserId>>,
-    users_for_key_query_cache: Arc<DashSet<OwnedUserId>>,
+    users_for_key_query: Arc<Mutex<UsersForKeyQuery>>,
     tracked_user_loading_lock: Arc<Mutex<()>>,
     tracked_users_loaded: Arc<AtomicBool>,
 }
 
+/// Record of the users that are waiting for a /keys/query.
+///
+/// To avoid races, we maintain a sequence number which is updated each time we
+/// receive an invalidation notification. We also record the sequence number at
+/// which each user was last invalidated. Then, we attach the current sequence
+/// number to each `/keys/query` request, and when we get the response we can
+/// tell if any users have been invalidated more recently than that request.
+#[derive(Debug)]
+struct UsersForKeyQuery {
+    /// The sequence number we will assign to the next addition to user_map
+    next_sequence_number: InvalidationSequenceNumber,
+
+    /// The users pending a lookup, together with the sequence number at which
+    /// they were added to the list
+    user_map: HashMap<OwnedUserId, InvalidationSequenceNumber>,
+}
+
+// We use wrapping arithmetic for the sequence numbers, to make sure we never
+// run out of numbers. (2**64 should be enough for anyone, but it's easy enough
+// just to make it wrap.)
+//
+// We use a *signed* counter so that we can compare values via a subtraction.
+// For example, suppose we've just overflowed from i64::MAX to i64::MIN.
+// (i64::MAX.wrapping_sub(i64::MIN)) is -1, which tells us that i64::MAX comes
+// before i64::MIN in the sequence.
+type InvalidationSequenceNumber = i64;
+
+impl UsersForKeyQuery {
+    /// Create a new, empty, `UsersForKeyQueryCache`
+    fn new() -> Self {
+        UsersForKeyQuery { next_sequence_number: 0, user_map: HashMap::new() }
+    }
+
+    /// Record a new user that requires a key query
+    fn insert_user(&mut self, user: &UserId) {
+        let seq = self.next_sequence_number;
+        trace!(?user, sequence_number = seq, "Flagging user for key query");
+        self.user_map.insert(user.to_owned(), seq);
+        self.next_sequence_number = self.next_sequence_number.wrapping_add(1);
+    }
+
+    /// Record that a user has received an update with the given sequence
+    /// number.
+    ///
+    /// If the sequence number is newer than the oldest invalidation for this
+    /// user, it is removed from the list of those needing an update.
+    ///
+    /// Returns true if the user is now up-to-date, else false
+    #[instrument(level = "trace", skip(self), fields(invalidation_sequence))]
```

Comment:
You missed a line in my code snippet:

```rust
Span::current().record("invalidation_sequence", invalidation_sequence);
```

The invocation of `fields(invalidation_sequence)` creates an empty field called `invalidation_sequence`, no value is attached to it, as we need to fetch the sequence number out of `self.user_map` we can't record it right away.

The `Span::current().record()` line will then attach the value to the field.

Reply:
gah, yes, thanks.